### PR TITLE
Identity/Similarity Strings for Alignment

### PR DESCRIPTION
**Pull request by bnbowman**
Addition of functions for the generation of similarity-strings or identity-strings for pairwise alignments or any two specified aligned rows, and associated demo/test-case.

As discussed on the mailing list: https://lists.fu-berlin.de/pipermail/seqan-dev/2015-January/msg00006.html

### CODE REVIEW

File: core/include/seqan/align/evaluate_alignment.h [hunk 1]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
     return computeAlignmentStats(stats, align, scoringScheme);
 }
 
+// ----------------------------------------------------------------------------
+// Function alignmentIdentityString()
+// ----------------------------------------------------------------------------
+
+/*!
+ * @fn Align#alignmentIdentityString()
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Make this interface global, by removing the "Align#" prefix. 
This is interface is not strictly bound to the Align object but should basically work with any alignment representing object. (For instance with the AlignmentGraph as well.)

File: core/include/seqan/align/evaluate_alignment.h [hunk 2]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+ * @headerfile <seqan/align.h>
+ * @brief Returns a CharString of the similarity of positions in an alignment
+ *
+ * @signature CharString alignmentIdentityString();
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Please add the complete signature using the described params below.
Optional arguments are put in brackets like: signature(mandatory [, optionalA] [, optionalB]);

Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: To add multiple signatures just add another signature to the @signature command.
Then you would have 2 signatures:
```
@signature alignmentIdentityString(align, ....);
           alignmentIdentityString(gapsH, gapsV, ...);

@param align ....
@param gapsH ....
@param gapsV ....
```

File: core/include/seqan/align/evaluate_alignment.h [hunk 3]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+ * @include demos/align/compute_match_string.cpp
+ *
+ * The output is as follows:
+ *
+ * @include demos/align/compute_match_string.cpp.stdout
+ */
+
+
+template <typename TGaps>
+String<char> alignmentIdentityString(TGaps & queryRow,
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: The function can be simplified using the std::transform interface.
Just add a function as follows:
``` cpp
template <typename TTarget, typename TSourceH, typename TSpecH, typename TSourceV, typename TSpecV, typename TBinaryOperator>
inline void transform(TTarget & target, 
                      Gaps<TSourceH, TGapsH> const & gapsH,
                      Gaps<TSourceV, TGapsV> const & gapsV,
                      TBinaryOperator const & binOperator)
{
    SEQAN_ASSERT_EQ(length(gapsH), length(gapsV));

    resize(target, length(gapsH), Exact());
    std::transform(begin(gapsH, Standard()), end(gapsH, Standard()), begin(gapsV, Standard()), begin(target, Standard()), binOperator);
}
```

At the moment we don't have the parallel version of transform implemented but we could easily switch later to use to the generic version (not before the release :)), which can then also be parallelised.

Then you simply create a functor that does the steps from line 310 to 326, except that your function-operator just returns the char that should be written to the target container. I added comments to the lines below.
The functor is then initialised with the chars for the different events.

File: core/include/seqan/align/evaluate_alignment.h [hunk 4]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+    TGapsIter itEnd1 = end(referenceRow);
+
+    // Counter for the position in the similarity string
+    unsigned pos = 0;
+
+    for (; it0 != itEnd0 && it1 != itEnd1; ++it0, ++it1, ++pos)
+    {
+        if (isGap(it0) || isGap(it1))
+        {
+            matchString[pos] = gapChar;
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Here in the functor return gapChar.

File: core/include/seqan/align/evaluate_alignment.h [hunk 5]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+        }
+        else
+        {
+            // Compute the alignment score and register in stats.
+            TAlphabet c0 = *it0, c1 = *it1;
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: These temporaries are not necessary. Just take the value of the iterator.

File: core/include/seqan/align/evaluate_alignment.h [hunk 6]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+            if (c0 == c1)
+            {
+                matchString[pos] = matchChar;
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Return matchChar

File: core/include/seqan/align/evaluate_alignment.h [hunk 7]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+            }
+            else
+            {
+                matchString[pos] = mismatchChar;
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: return mismatchChar.

File: core/include/seqan/align/evaluate_alignment.h [hunk 8]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+        }
+    }
+    SEQAN_ASSERT(it0 == itEnd0);
+    SEQAN_ASSERT(it1 == itEnd1);
+
+    return matchString;
+}
+
+template <typename TGaps>
+String<char> alignmentIdentityString(TGaps const & queryRow,
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Please inline all wrapper functions.

File: core/include/seqan/align/evaluate_alignment.h [hunk 9]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+    TGapsIter itEnd0 = end(queryRow);
+    TGapsIter it1 = begin(referenceRow);
+    TGapsIter itEnd1 = end(referenceRow);
+
+    // Counter for the position in the similarity string
+    unsigned pos = 0;
+
+    for (; it0 != itEnd0 && it1 != itEnd1; ++it0, ++it1, ++pos)
+    {
+        if (isGap(it0) || isGap(it1))
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: From here this part goes into the functor.

File: core/include/seqan/align/evaluate_alignment.h [hunk 10]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+    TGapsIter it0 = begin(queryRow);
+    TGapsIter itEnd0 = end(queryRow);
+    TGapsIter it1 = begin(referenceRow);
+    TGapsIter itEnd1 = end(referenceRow);
+
+    // Counter for the position in the similarity string
+    unsigned pos = 0;
+
+    for (; it0 != itEnd0 && it1 != itEnd1; ++it0, ++it1, ++pos)
+    {
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Remove everything from above and only keep the matchString definition. 
Also this function then gets as third parameter the functor not the chars to write the string.
These are then stored within the functor.
Then call transform() as defined in the comment above and return matchString.

File: core/include/seqan/align/evaluate_alignment.h [hunk 11]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+    SEQAN_ASSERT(it1 == itEnd1);
+
+    return matchString;
+}
+
+template <typename TGaps>
+String<char> alignmentIdentityString(TGaps const & queryRow,
+                                     TGaps const & referenceRow)
+{
+    return alignmentIdentityString(queryRow, referenceRow, '|', '*', ' ');
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Calls this function with the functor instead, like
```cpp
    return alignmentIdentityString(queryRow, referenceRow, IdentityStringGenerator('|', '*', ' ' ));
```

File: core/include/seqan/align/evaluate_alignment.h [hunk 12]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+                                     char const gapChar)
+{
+    // Check that the supplied alignment is pairwise
+    SEQAN_ASSERT_EQ_MSG(length(rows(align)), 2u, "Only works with pairwise alignments.");
+
+    typedef Align<TAlphabet, TAlignSpec> const TAlign;
+    typedef typename Row<TAlign>::Type TGaps;
+
+    TGaps queryRow = row(align, 0);
+    TGaps referenceRow = row(align, 1);
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Use TGaps & instead. Avoids copying.

File: core/include/seqan/align/evaluate_alignment.h [hunk 13]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+
+    return alignmentIdentityString(queryRow, referenceRow, matchChar, mismatchChar, gapChar);
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Same as call above.

File: core/include/seqan/align/evaluate_alignment.h [hunk 14]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+}
+
+template <typename TSource, typename TAlignSpec>
+String<char> alignmentIdentityString(Align<TSource, TAlignSpec> const & align)
+{
+    return alignmentIdentityString(align, '|', '*', ' ');
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Same as above.

File: core/include/seqan/align/evaluate_alignment.h [hunk 15]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+// ----------------------------------------------------------------------------
+// Function alignmentSimilarityString()
+// ----------------------------------------------------------------------------
+
+/*!
+ * @fn Align#alignmentSimilarityString()
+ * @headerfile <seqan/align.h>
+ * @brief Returns a CharString of the similarity of positions in an alignment
+ *
+ * @signature CharString computeAlignmentSimilarityString();
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Pease add the full signatures.

File: core/include/seqan/align/evaluate_alignment.h [hunk 16]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+ *
+ * @include demos/align/compute_match_string.cpp
+ *
+ * The output is as follows:
+ *
+ * @include demos/align/compute_match_string.cpp.stdout
+ */
+
+template <typename TGaps, typename TScoreVal, typename TScoreSpec>
+String<char> alignmentSimilarityString(TGaps & queryRow,
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Now you can reuse the interface from above to do the same here.
Maybe recall this interface (above) to _transformAlignment(Gaps, Gaps, BinaryOperator).
This should be then an internal interface, that's why the underscore at the beginning.

File: core/include/seqan/align/evaluate_alignment.h [hunk 17]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+    TGapsIter itEnd0 = it0 + alignLength;
+    TGapsIter it1 = begin(referenceRow);
+    TGapsIter itEnd1 = it1 + alignLength;
+
+    // Counter for the position in the similarity string
+    unsigned pos = 0;
+
+    for (; it0 != itEnd0 && it1 != itEnd1; ++it0, ++it1, ++pos)
+    {
+        if (isGap(it0) || isGap(it1))
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Then only write another functor, maybe called 
```cpp
template <typename TTarget>
struct SimilarityStringGenerator
{
    typename Value<TTarget>::Type TTargetValue;
    
    TTargetValue  matchChar;
    ...

    template <typename TValueH, typename TValueV>
    TTargetValue operator()(TValueH valH, TValueV valV)
    {
        // The logic from line 424 to 443.
    }
};

Similar you write the functor above.

File: core/include/seqan/align/evaluate_alignment.h [hunk 18]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+    }
+    SEQAN_ASSERT(it0 == itEnd0);
+    SEQAN_ASSERT(it1 == itEnd1);
+
+    return matchString;
+}
+
+
+template <typename TGaps, typename TScoreVal, typename TScoreSpec>
+String<char> alignmentSimilarityString(TGaps const & queryRow,
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Please inline all functions.

File: core/include/seqan/align/evaluate_alignment.h [hunk 19]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+                                       TGaps const & referenceRow,
+                                       Score<TScoreVal, TScoreSpec> const & scoringScheme)
+{
+    return alignmentSimilarityString(queryRow, referenceRow, scoringScheme,
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Use the functor as described above instead.

File: core/include/seqan/align/evaluate_alignment.h [hunk 20]
```diff
@@ -252,6 +252,241 @@ TScoreVal computeAlignmentStats(Align<TGaps, TAlignSpec> const & align,
...
+}
+
+template <typename TSource, typename TAlignSpec, typename TScoreVal, typename TScoreSpec>
+String<char> alignmentSimilarityString(Align<TSource, TAlignSpec> const & align,
+                                       Score<TScoreVal, TScoreSpec> const & scoringScheme)
+{
+    return alignmentSimilarityString(align, scoringScheme, '|', ':', '.', ' ');
+}
+
+
```
Comment (rrahn) [inline core/include/seqan/align/evaluate_alignment.h]: Remove superfluous empty line.

**Comment by rrahn** [comment]
@bnbowman what is the state of this PR.
In the meantime there have been some changes to this file. Could you please rebase your changes first?

**Comment by h-2** [comment]
@bnbowman  Since there have been no updates to this PR in nearly a year, it will be closed if it's not updated for 2.3.0.

**Pull request closed without merging**